README.md
<!-- [![build_master](https://img.shields.io/travis/pleiszenburg/wenv-kernel/master.svg?style=flat-square "Build Status: master / release")](https://travis-ci.org/pleiszenburg/wenv-kernel) -->
<!-- [![build_develop](https://img.shields.io/travis/pleiszenburg/wenv-kernel/develop.svg?style=flat-square "Build Status: development branch")](https://travis-ci.org/pleiszenburg/wenv-kernel) -->
[![license](https://img.shields.io/pypi/l/wenvkernel.svg?style=flat-square "GNU Lesser General Public License v2.1")](https://github.com/pleiszenburg/wenv-kernel/blob/master/LICENSE)
[![status](https://img.shields.io/pypi/status/wenvkernel.svg?style=flat-square "Project Development Status")](https://github.com/pleiszenburg/wenv-kernel/issues)
[![pypi_version](https://img.shields.io/pypi/v/wenvkernel.svg?style=flat-square "Project Development Status")](https://pypi.python.org/pypi/wenvkernel)
[![pypi_versions](https://img.shields.io/pypi/pyversions/wenvkernel.svg?style=flat-square "Available on PyPi - the Python Package Index")](https://pypi.python.org/pypi/wenvkernel)

![wenv-kernel](http://www.pleiszenburg.de/wenv-kernel_logo.png)

## Synopsis

**wenv-kernel** is a **Python package** (currently in development **status 4/beta**). It allows to **run Python on top of Wine as a Jupyter kernel** on Linux, MacOS or BSD. It uses **[wenv](https://github.com/pleiszenburg/wenv)**.

## Prerequisites

A working installation of `wenv`, see [installation instructions](https://wenv.readthedocs.io/en/latest/installation.html) in its documentation. If not present, `wenv-kernel` will try to install and configure `wenv` automatically. It is assumed that *Wine* is present.

## Installation

```bash
pip install wenvkernel
python -m wenvkernel.install
```

After installing the package with `pip`, you must not forget to run the kernel install module once as shown above.

## Examples

| launcher | platform | ipc |
|:--------:|:--------:|:---:|
| ![launcher](https://github.com/pleiszenburg/wenv-kernel/blob/master/docs/launcher.png?raw=true "launcher") | ![platform](https://github.com/pleiszenburg/wenv-kernel/blob/master/docs/platform.png?raw=true "platform") | ![ipc](https://github.com/pleiszenburg/wenv-kernel/blob/master/docs/ipc.png?raw=true "ipc") |

See [examples](https://github.com/pleiszenburg/wenv-kernel/blob/master/examples) for example Jupyter notebooks.

## Need help?

Feel free to post questions in the [GitHub issue tracker](https://github.com/pleiszenburg/wenv-kernel/labels/question) of this project.

## Bugs & issues

- Report bugs in *wenv-kernel* here: [wenv-kernel GitHub issue tracker](https://github.com/pleiszenburg/wenv-kernel/issues)
- Report bugs in *wenv* here: [wenv GitHub issue tracker](https://github.com/pleiszenburg/wenv/issues)
- Report bugs in *Wine* here: [WineHQ Bug Tracking System](https://bugs.winehq.org/)

## Miscellaneous

- [Authors](https://github.com/pleiszenburg/wenv-kernel/blob/master/AUTHORS.md)
- [Change log (current)](https://github.com/pleiszenburg/wenv-kernel/blob/develop/CHANGES.md) (changes in development branch since last release)
- [Change log (past)](https://github.com/pleiszenburg/wenv-kernel/blob/master/CHANGES.md) (release history)
- [Contributing](https://github.com/pleiszenburg/wenv-kernel/blob/master/CONTRIBUTING.md) (**Contributions are highly welcomed!**)
- [License](https://github.com/pleiszenburg/wenv-kernel/blob/master/LICENSE) (**LGPL v2.1**)


CHANGES.md
# Changes

## 0.0.2 (2020-XX-XX)

- (TBD)

## 0.0.1 (2020-07-20)

- Initial release


tests/test_32bit.py
# -*- coding: utf-8 -*-

"""

WENV-KERNEL
Jupyter kernel for Python on Wine
https://github.com/pleiszenburg/wenv-kernel

    tests/test_32bit.py: Testing 32 bit kernel

    Copyright (C) 2017-2020 Sebastian M. Ernst <ernst@pleiszenburg.de>

<LICENSE_BLOCK>
The contents of this file are subject to the GNU Lesser General Public License
Version 2.1 ("LGPL" or "License"). You may not use this file except in
compliance with the License. You may obtain a copy of the License at
https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt
https://github.com/pleiszenburg/wenv-kernel/blob/master/LICENSE

Software distributed under the License is distributed on an "AS IS" basis,
WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for the
specific language governing rights and limitations under the License.
</LICENSE_BLOCK>

"""

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# IMPORT
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

import unittest
import jupyter_kernel_test

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# TESTS
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++


class MyKernelTests(jupyter_kernel_test.KernelTests):
    # Required --------------------------------------

    # The name identifying an installed kernel to run the tests against
    kernel_name = "wenv_python3_32bit"

    # language_info.name in a kernel_info_reply should match this
    language_name = "python"

    # Optional --------------------------------------

    # Code in the kernel's language to write "hello, world" to stdout
    code_hello_world = "print('hello, world')"

    # Pager: code that should display something (anything) in the pager
    code_page_something = "help(print)"

    # Samples of code which generate a result value (ie, some text
    # displayed as Out[n])
    code_execute_result = [{"code": "6*7", "result": "42"}]

    # Samples of code which should generate a rich display output, and
    # the expected MIME type
    # code_display_data = [
    # 	{'code': 'show_image()', 'mime': 'image/png'}
    # ]

    # You can also write extra tests. We recommend putting your kernel name
    # in the method name, to avoid clashing with any tests that
    # jupyter_kernel_test adds in the future.
    # def test_wenv_python3_32bit_stderr(self):
    # 	reply, output_msgs = self.execute_helper(code='print_err "oops"')
    # 	self.assertEqual(output_msgs[0].header['msg_type'], 'stream')
    # 	self.assertEqual(output_msgs[0].content['name'], 'stderr')
    # 	self.assertEqual(output_msgs[0].content['text'], 'oops\n')


tests/test_64bit.py
# -*- coding: utf-8 -*-

"""

WENV-KERNEL
Jupyter kernel for Python on Wine
https://github.com/pleiszenburg/wenv-kernel

    tests/test_64bit.py: Testing 64 bit kernel

    Copyright (C) 2017-2020 Sebastian M. Ernst <ernst@pleiszenburg.de>

<LICENSE_BLOCK>
The contents of this file are subject to the GNU Lesser General Public License
Version 2.1 ("LGPL" or "License"). You may not use this file except in
compliance with the License. You may obtain a copy of the License at
https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt
https://github.com/pleiszenburg/wenv-kernel/blob/master/LICENSE

Software distributed under the License is distributed on an "AS IS" basis,
WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for the
specific language governing rights and limitations under the License.
</LICENSE_BLOCK>

"""

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# IMPORT
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

import unittest
import jupyter_kernel_test

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# TESTS
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++


class MyKernelTests(jupyter_kernel_test.KernelTests):
    # Required --------------------------------------

    # The name identifying an installed kernel to run the tests against
    kernel_name = "wenv_python3_64bit"

    # language_info.name in a kernel_info_reply should match this
    language_name = "python"

    # Optional --------------------------------------

    # Code in the kernel's language to write "hello, world" to stdout
    code_hello_world = "print('hello, world')"

    # Pager: code that should display something (anything) in the pager
    code_page_something = "help(print)"

    # Samples of code which generate a result value (ie, some text
    # displayed as Out[n])
    code_execute_result = [{"code": "6*7", "result": "42"}]

    # Samples of code which should generate a rich display output, and
    # the expected MIME type
    # code_display_data = [
    # 	{'code': 'show_image()', 'mime': 'image/png'}
    # ]

    # You can also write extra tests. We recommend putting your kernel name
    # in the method name, to avoid clashing with any tests that
    # jupyter_kernel_test adds in the future.
    # def test_wenv_python3_32bit_stderr(self):
    # 	reply, output_msgs = self.execute_helper(code='print_err "oops"')
    # 	self.assertEqual(output_msgs[0].header['msg_type'], 'stream')
    # 	self.assertEqual(output_msgs[0].content['name'], 'stderr')
    # 	self.assertEqual(output_msgs[0].content['text'], 'oops\n')


makefile.py
# -*- coding: utf-8 -*-

"""

WENV-KERNEL
Jupyter kernel for Python on Wine
https://github.com/pleiszenburg/wenv-kernel

	makefile.py: Packaging

	Copyright (C) 2017-2020 Sebastian M. Ernst <ernst@pleiszenburg.de>

<LICENSE_BLOCK>
The contents of this file are subject to the GNU Lesser General Public License
Version 2.1 ("LGPL" or "License"). You may not use this file except in
compliance with the License. You may obtain a copy of the License at
https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt
https://github.com/pleiszenburg/wenv-kernel/blob/master/LICENSE

Software distributed under the License is distributed on an "AS IS" basis,
WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for the
specific language governing rights and limitations under the License.
</LICENSE_BLOCK>

"""

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# IMPORT
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

import base64
import os

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# CONST
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

LOGO_PY = os.path.join("src", "wenvkernel", "_logos.py")
RESOURCES_FLD = "res"
LOGO_FN = "logo-{SIZE:d}x{SIZE:d}_{BITS:d}.png"
SIZES = (32, 64)
BITS = (32, 64)
LINE_LEN = 72

TEMPLATE_FILE = """# -*- coding: utf-8 -*-
# This file is part of WENV-KERNEL.
# Do not change it manually. Changes will be lost on next re-build.

LOGO_SIZES = (32, 64)

LOGOS = {{

{CONTENT}

}}
"""
TEMPLATE_VAR = '({SIZE:d}, {BITS:d}): """\n{IMAGE:s}\n""",'

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# ROUTINES
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++


def resources():

    images = {(bits, size): _get_image(bits, size) for size in SIZES for bits in BITS}

    content = [
        TEMPLATE_VAR.format(BITS=bits, SIZE=size, IMAGE=image)
        for (bits, size), image in images.items()
    ]

    with open(LOGO_PY, "w", encoding="utf-8") as f:
        f.write(TEMPLATE_FILE.format(CONTENT="\n\n".join(content)))


def _get_image(bits, size):

    fn = LOGO_FN.format(BITS=bits, SIZE=size)
    with open(os.path.join(RESOURCES_FLD, fn), "rb") as f:
        data = f.read()

    data = base64.b64encode(data).decode("utf-8")

    return "\n".join(
        data[pos : (pos + LINE_LEN)] for pos in range(0, len(data), LINE_LEN)
    )


src/wenvkernel/install.py
# -*- coding: utf-8 -*-

"""

WENV-KERNEL
Jupyter kernel for Python on Wine
https://github.com/pleiszenburg/wenv-kernel

    src/wenvkernel/install.py: User install entry point

    Copyright (C) 2017-2020 Sebastian M. Ernst <ernst@pleiszenburg.de>

<LICENSE_BLOCK>
The contents of this file are subject to the GNU Lesser General Public License
Version 2.1 ("LGPL" or "License"). You may not use this file except in
compliance with the License. You may obtain a copy of the License at
https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt
https://github.com/pleiszenburg/wenv-kernel/blob/master/LICENSE

Software distributed under the License is distributed on an "AS IS" basis,
WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for the
specific language governing rights and limitations under the License.
</LICENSE_BLOCK>

"""

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# IMPORT
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

from ._core import setup_kernels as _setup_kernels

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# ENTRY POINT
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

if __name__ == "__main__":
    _setup_kernels()


src/wenvkernel/__init__.py
# -*- coding: utf-8 -*-

"""

WENV-KERNEL
Jupyter kernel for Python on Wine
https://github.com/pleiszenburg/wenv-kernel

    src/wenvkernel/__init__.py: Package init file

    Copyright (C) 2017-2020 Sebastian M. Ernst <ernst@pleiszenburg.de>

<LICENSE_BLOCK>
The contents of this file are subject to the GNU Lesser General Public License
Version 2.1 ("LGPL" or "License"). You may not use this file except in
compliance with the License. You may obtain a copy of the License at
https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt
https://github.com/pleiszenburg/wenv-kernel/blob/master/LICENSE

Software distributed under the License is distributed on an "AS IS" basis,
WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for the
specific language governing rights and limitations under the License.
</LICENSE_BLOCK>

"""

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# IMPORT / EXPORT
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

from ._core import setup_kernels as _setup_kernels, setup_kernel as _setup_kernel


src/wenvkernel/_core.py
# -*- coding: utf-8 -*-

"""

WENV-KERNEL
Jupyter kernel for Python on Wine
https://github.com/pleiszenburg/wenv-kernel

    src/wenvkernel/_core.py: Package core

    Copyright (C) 2017-2020 Sebastian M. Ernst <ernst@pleiszenburg.de>

<LICENSE_BLOCK>
The contents of this file are subject to the GNU Lesser General Public License
Version 2.1 ("LGPL" or "License"). You may not use this file except in
compliance with the License. You may obtain a copy of the License at
https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt
https://github.com/pleiszenburg/wenv-kernel/blob/master/LICENSE

Software distributed under the License is distributed on an "AS IS" basis,
WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for the
specific language governing rights and limitations under the License.
</LICENSE_BLOCK>

"""

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# CONST
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

DEFAULT_TIMEOUT = 600
LOGO_FN = "logo-{SIZE:d}x{SIZE:d}.png"
KERNEL_FN = "kernel.json"
KERNEL_NAME = "wenv_python3"
KERNEL_TEMPLATE = {
    "argv": ["wenv", "python", "-m", "ipykernel_launcher", "-f", "{connection_file}"],
    "display_name": "wenv python",
    "language": "python",
}

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# IMPORT
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

import base64
import json
import os
import subprocess
import sys

from ._logos import LOGOS, LOGO_SIZES

# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# IMPORT
# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++


def setup_kernels():

    for bits in (32, 64):
        setup_kernel(bits)


def setup_kernel(bits):

    kernel_name = _get_name(bits)
    kernel_path = _get_path(kernel_name)
    _ensure_path(kernel_path)
    _write_config(kernel_path, bits)
    for size in LOGO_SIZES:
        _write_image(kernel_path, bits, size)
    _wenv_init(bits)
    _wenv_pip(bits, "install", "ipykernel")


def _ensure_path(kernel_path):

    os.makedirs(kernel_path, exist_ok=True)


def _get_name(bits):

    return KERNEL_NAME + "_{BITS:d}bit".format(BITS=bits)


def _get_path(kernel_name):

    return os.path.join(sys.prefix, "share", "jupyter", "kernels", kernel_name)


def _run_process(cmd_list, env=None, timeout=DEFAULT_TIMEOUT):

    if env is None:
        env = {}
    envvar_dict = {k: os.environ[k] for k in os.environ.keys()}
    envvar_dict.update(env)

    proc = subprocess.Popen(
        cmd_list, stdout=subprocess.PIPE, stderr=subprocess.PIPE, env=envvar_dict
    )
    try:
        outs, errs = proc.communicate(timeout=timeout)
    except subprocess.TimeoutExpired:
        os.kill(proc.pid, signal.SIGINT)
        outs, errs = proc.communicate()

    return (outs.decode("utf-8"), errs.decode("utf-8"), proc.returncode)


def _wenv_init(bits):

    out, err, code = _run_process(
        ["wenv", "init"], env={"WENV_ARCH": "win{BITS:d}".format(BITS=bits)}
    )
    if code != 0:
        raise SystemError("wenv init failed", out, err)


def _wenv_pip(bits, *args):

    out, err, code = _run_process(
        ["wenv", "pip"] + list(args), env={"WENV_ARCH": "win{BITS:d}".format(BITS=bits)}
    )
    if code != 0:
        raise SystemError("wenv pip failed", out, err)


def _write_config(kernel_path, bits):

    kernel = KERNEL_TEMPLATE.copy()
    kernel["display_name"] += " {BITS:d}bit".format(BITS=bits)
    kernel["env"] = {"WENV_ARCH": "win{BITS:d}".format(BITS=bits)}

    with open(os.path.join(kernel_path, KERNEL_FN), "w", encoding="utf-8") as f:
        f.write(json.dumps(kernel, indent=4, sort_keys=False))


def _write_image(kernel_path, bits, size):

    with open(os.path.join(kernel_path, LOGO_FN.format(SIZE=size)), "wb") as f:
        f.write(base64.b64decode(LOGOS[(size, bits)].strip().replace("\n", "")))
